src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -382,13 +387,14 @@ private SortedMap<PathFragment, ActionInput> buildOutputDirMap(Spawn spawn) {
     return outputDirMap;
   }
 
-  private MerkleTree buildInputMerkleTree(Spawn spawn, SpawnExecutionContext context)
+  private MerkleTree buildInputMerkleTree(
+      Spawn spawn, SpawnExecutionContext context, ToolSignature toolSignature)
       throws IOException, ForbiddenActionInputException {
     // Add output directories to inputs so that they are created as empty directories by the
     // executor. The spec only requires the executor to create the parent directory of an output
     // directory, which differs from the behavior of both local and sandboxed execution.
     SortedMap<PathFragment, ActionInput> outputDirMap = buildOutputDirMap(spawn);
-    if (remoteOptions.remoteMerkleTreeCache) {
+    if (remoteOptions.remoteMerkleTreeCache && toolSignature == null) {
```

Comment:
This change was developed before the merkle tree cache was merged. Performance benefits were realized for the intended usecases. So, when the merkle tree cache came along, it didn't seem necessary to integrate tightly with it. I think our benchmarks didn't benefit from the merkle tree cache anyway because, for example, `Javac` actions normally don't have massive input trees.

No doubt this could be improved but one has to do the thinking to make sure if inputs move in and out of the tool set, the merkle tree is invalidated properly.

Reply:
Ok, that can be a follow-up item. Just mark it as such here, otherwise this `if` is very confusing.